In [1]:
# Cai dat va kiem tra moi truong lam viec
import torch
import numpy as np
import random

torch.manual_seed(42)
np.random.seed(42)
random.seed(42)

print("PyTorch version:", torch.__version__)
print("CUDA available :", torch.cuda.is_available())

PyTorch version: 2.14.0+cu130
CUDA available : False


### Bài 1. Tạo Tensor cơ bản
Tạo một Tensor số nguyên kích thước 2 × 3 và hiển thị kết quả.

In [2]:
import torch

a = torch.tensor([[1, 2, 3],
                  [4, 5, 6]], dtype=torch.int32)
print(a)

tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)


**Yêu cầu:** Cho biết kích thước, số chiều và kiểu dữ liệu của Tensor a.

In [3]:
print("Kich thuoc (shape)      :", a.shape)
print("So chieu (ndim)         :", a.ndim)
print("Kieu du lieu (dtype)    :", a.dtype)
print("Tong so phan tu (numel) :", a.numel())

Kich thuoc (shape)      : torch.Size([2, 3])
So chieu (ndim)         : 2
Kieu du lieu (dtype)    : torch.int32
Tong so phan tu (numel) : 6


**Nhận xét:** Tensor `a` có kích thước `torch.Size([2, 3])` (2 hàng, 3 cột), số chiều `ndim = 2`
(tensor 2 chiều - ma trận), kiểu dữ liệu `torch.int32` (số nguyên 32 bit) và tổng cộng 6 phần tử.


### Bài 2. Tạo các Tensor thường dùng
Thực hành tạo Tensor ngẫu nhiên, Tensor toàn số 0, toàn số 1 và ma trận đơn vị.

In [4]:
import torch

a = torch.rand(3, 4)    # phan phoi deu U(0, 1)
b = torch.randn(3, 4)   # phan phoi chuan N(0, 1)
c = torch.zeros(2, 3)   # toan so 0
d = torch.ones(2, 3)    # toan so 1
e = torch.eye(4)        # ma tran don vi 4x4

print("a = torch.rand(3,4):");  print(a)
print("\nb = torch.randn(3,4):"); print(b)
print("\nc = torch.zeros(2,3):"); print(c)
print("\nd = torch.ones(2,3):");  print(d)
print("\ne = torch.eye(4):");     print(e)

a = torch.rand(3,4):
tensor([[0.8823, 0.9150, 0.3829, 0.9593],
        [0.3904, 0.6009, 0.2566, 0.7936],
        [0.9408, 0.1332, 0.9346, 0.5936]])

b = torch.randn(3,4):
tensor([[ 2.2082, -0.6380,  0.4617,  0.2674],
        [ 0.5349,  0.8094,  1.1103, -1.6898],
        [-0.9890,  0.9580,  1.3221,  0.8172]])

c = torch.zeros(2,3):
tensor([[0., 0., 0.],
        [0., 0., 0.]])

d = torch.ones(2,3):
tensor([[1., 1., 1.],
        [1., 1., 1.]])

e = torch.eye(4):
tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])


**Yêu cầu 1:** Nhận xét sự khác nhau giữa `torch.rand()` và `torch.randn()`.

In [5]:
r1 = torch.rand(10000)
r2 = torch.randn(10000)
print("torch.rand : min=%.4f  max=%.4f  mean=%.4f  std=%.4f" % (r1.min(), r1.max(), r1.mean(), r1.std()))
print("torch.randn: min=%.4f  max=%.4f  mean=%.4f  std=%.4f" % (r2.min(), r2.max(), r2.mean(), r2.std()))

torch.rand : min=0.0001  max=0.9999  mean=0.5006  std=0.2888
torch.randn: min=-4.5905  max=3.7069  mean=0.0055  std=0.9970


**Nhận xét:**
- `torch.rand()` sinh số theo **phân phối đều** trên nửa khoảng [0, 1): mọi giá trị đều nằm trong [0,1),
  trung bình xấp xỉ 0.5, không có giá trị âm.
- `torch.randn()` sinh số theo **phân phối chuẩn (Gauss) N(0, 1)**: giá trị có thể âm hoặc dương,
  không bị chặn, trung bình xấp xỉ 0 và độ lệch chuẩn xấp xỉ 1.
- Trong học sâu, `randn` thường được dùng để khởi tạo trọng số vì phân phối đối xứng quanh 0.

**Yêu cầu 2:** Tạo thêm một Tensor 5 × 5 chứa toàn giá trị 10.

In [6]:
f = torch.full((5, 5), 10)          # cach 1: torch.full
g = torch.ones(5, 5) * 10           # cach 2: nhan vo huong
print("torch.full((5,5), 10):"); print(f)
print("\ntorch.ones(5,5) * 10:"); print(g)

torch.full((5,5), 10):
tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])

torch.ones(5,5) * 10:
tensor([[10., 10., 10., 10., 10.],
        [10., 10., 10., 10., 10.],
        [10., 10., 10., 10., 10.],
        [10., 10., 10., 10., 10.],
        [10., 10., 10., 10., 10.]])



### Bài 3. Kiểm tra thông tin Tensor
Sử dụng `shape`, `size()`, `ndim` và `dtype` để kiểm tra các thuộc tính của Tensor.

In [7]:
import torch

a = torch.zeros(2, 3, 4)
print("Tensor:")
print(a)
print("\nShape              :", a.shape)
print("Size               :", a.size())
print("Number of dimensions:", a.ndim)
print("Data type          :", a.dtype)
print("Number of elements :", a.numel())

Tensor:
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])

Shape              : torch.Size([2, 3, 4])
Size               : torch.Size([2, 3, 4])
Number of dimensions: 3
Data type          : torch.float32
Number of elements : 24


**Yêu cầu:** Giải thích ý nghĩa của `torch.Size([2, 3, 4])`.

**Trả lời:** `torch.Size([2, 3, 4])` cho biết tensor có **3 chiều**:
- Chiều 0 (dim=0) có 2 phần tử → 2 "khối" (block/batch).
- Chiều 1 (dim=1) có 3 phần tử → mỗi khối là ma trận 3 hàng.
- Chiều 2 (dim=2) có 4 phần tử → mỗi hàng có 4 cột.

Tổng số phần tử = 2 × 3 × 4 = 24. `torch.Size` thực chất là một lớp con của `tuple`,
nên có thể truy xuất như tuple: `a.shape[0] = 2`. Ngoài ra `a.shape` và `a.size()` cho kết quả giống hệt nhau,
chỉ khác `shape` là thuộc tính còn `size()` là phương thức.


### Bài 4. Slicing và Indexing
Truy xuất một phần dữ liệu từ Tensor nhiều chiều.

In [8]:
import torch

a = torch.randn(3, 4, 5)
print("Original tensor:")
print(a)
print("\nTensor a[1]:")
print(a[1])
print("\nTensor a[1:, 2:4]:")
print(a[1:, 2:4])

Original tensor:
tensor([[[ 0.3605, -0.8189,  1.0898,  1.2362, -0.8152],
         [ 0.3237,  0.8048,  0.4700, -1.7766,  0.2693],
         [ 0.8466,  0.7630,  0.3405, -1.7775, -0.2623],
         [-1.4898, -0.8240,  0.4119,  0.8887, -0.0233]],

        [[-0.1569,  0.1672,  0.2644, -0.3617,  0.9134],
         [ 1.0054,  0.9127, -0.8800, -0.9963, -1.7399],
         [ 0.3507, -0.5245,  1.6081,  1.0534, -0.7705],
         [ 0.7094,  1.7103, -0.1474,  1.3686,  0.4873]],

        [[ 0.5693, -0.8539, -0.1527, -0.3782, -1.2409],
         [ 0.1008,  0.6087,  1.0949, -0.6739,  1.1621],
         [ 1.1202, -0.1299,  1.4179,  1.2474, -0.3297],
         [ 0.6028,  0.6084,  0.3874,  1.4482,  0.3966]]])

Tensor a[1]:
tensor([[-0.1569,  0.1672,  0.2644, -0.3617,  0.9134],
        [ 1.0054,  0.9127, -0.8800, -0.9963, -1.7399],
        [ 0.3507, -0.5245,  1.6081,  1.0534, -0.7705],
        [ 0.7094,  1.7103, -0.1474,  1.3686,  0.4873]])

Tensor a[1:, 2:4]:
tensor([[[ 0.3507, -0.5245,  1.6081,  1.0534, -0.7

**Yêu cầu 1:** Cho biết shape của `a[1]` và `a[1:, 2:4]`.

In [9]:
print("shape cua a          :", a.shape)
print("shape cua a[1]       :", a[1].shape)
print("shape cua a[1:, 2:4] :", a[1:, 2:4].shape)

shape cua a          : torch.Size([3, 4, 5])
shape cua a[1]       : torch.Size([4, 5])
shape cua a[1:, 2:4] : torch.Size([2, 2, 5])


**Nhận xét:**
- `a[1]` là **indexing** trên chiều 0 → chiều này bị "nuốt" mất, kết quả có shape `[4, 5]` (giảm 1 chiều).
- `a[1:, 2:4]` là **slicing**: chiều 0 lấy từ chỉ số 1 đến hết → còn 2 phần tử; chiều 1 lấy chỉ số 2 và 3 →
  còn 2 phần tử; chiều 2 giữ nguyên 5 → shape `[2, 2, 5]` (vẫn 3 chiều).
- Điểm mấu chốt: **index bằng số nguyên làm giảm số chiều, còn slice `:` thì giữ nguyên số chiều.**

**Yêu cầu 2:** Thử lấy hai phần tử cuối cùng trên chiều thứ ba.

In [10]:
last2 = a[:, :, -2:]
print("a[:, :, -2:] shape:", last2.shape)
print(last2)

a[:, :, -2:] shape: torch.Size([3, 4, 2])
tensor([[[ 1.2362, -0.8152],
         [-1.7766,  0.2693],
         [-1.7775, -0.2623],
         [ 0.8887, -0.0233]],

        [[-0.3617,  0.9134],
         [-0.9963, -1.7399],
         [ 1.0534, -0.7705],
         [ 1.3686,  0.4873]],

        [[-0.3782, -1.2409],
         [-0.6739,  1.1621],
         [ 1.2474, -0.3297],
         [ 1.4482,  0.3966]]])


**Nhận xét:** `a[:, :, -2:]` giữ nguyên chiều 0 và chiều 1, chỉ lấy 2 phần tử cuối của chiều 2,
nên shape từ `[3, 4, 5]` trở thành `[3, 4, 2]`.


### Bài 5. Reshape và Flatten
Thay đổi hình dạng Tensor mà không làm thay đổi số phần tử.

In [11]:
import torch

a = torch.randn(3, 4)
print("Original:")
print(a)
print("\nFlatten:")
print(a.ravel())
print("\nReshape to 3 x 2 x 2:")
print(a.reshape(3, 2, 2))

Original:
tensor([[-0.1367, -1.3296,  0.6277, -1.2770],
        [-0.2235, -0.8259, -1.8966,  1.7000],
        [-0.2159,  0.2072,  0.7678,  0.6904]])

Flatten:
tensor([-0.1367, -1.3296,  0.6277, -1.2770, -0.2235, -0.8259, -1.8966,  1.7000,
        -0.2159,  0.2072,  0.7678,  0.6904])

Reshape to 3 x 2 x 2:
tensor([[[-0.1367, -1.3296],
         [ 0.6277, -1.2770]],

        [[-0.2235, -0.8259],
         [-1.8966,  1.7000]],

        [[-0.2159,  0.2072],
         [ 0.7678,  0.6904]]])


**Yêu cầu 1:** Kiểm tra số phần tử trước và sau khi reshape.

In [12]:
b = a.reshape(3, 2, 2)
print("Truoc reshape: shape =", a.shape, "| so phan tu =", a.numel())
print("Sau  reshape : shape =", b.shape, "| so phan tu =", b.numel())
print("So phan tu co bang nhau khong?", a.numel() == b.numel())

Truoc reshape: shape = torch.Size([3, 4]) | so phan tu = 12
Sau  reshape : shape = torch.Size([3, 2, 2]) | so phan tu = 12
So phan tu co bang nhau khong? True


**Yêu cầu 2:** Thử reshape Tensor thành kích thước 2 × 6.

In [13]:
c = a.reshape(2, 6)
print("a.reshape(2, 6):")
print(c)
print("shape:", c.shape, "| numel:", c.numel())

# Dung -1 de PyTorch tu suy ra so chieu con lai
print("\na.reshape(-1, 6).shape =", a.reshape(-1, 6).shape)
print("a.reshape(2, -1).shape  =", a.reshape(2, -1).shape)

# Thu reshape sai kich thuoc de thay loi
try:
    a.reshape(5, 3)
except RuntimeError as err:
    print("\nLoi khi reshape(5,3):", err)

a.reshape(2, 6):
tensor([[-0.1367, -1.3296,  0.6277, -1.2770, -0.2235, -0.8259],
        [-1.8966,  1.7000, -0.2159,  0.2072,  0.7678,  0.6904]])
shape: torch.Size([2, 6]) | numel: 12

a.reshape(-1, 6).shape = torch.Size([2, 6])
a.reshape(2, -1).shape  = torch.Size([2, 6])

Loi khi reshape(5,3): shape '[5, 3]' is invalid for input of size 12


**Nhận xét:**
- `ravel()` (hoặc `flatten()`) trải tensor thành vector 1 chiều 12 phần tử.
- Reshape chỉ thay đổi **cách diễn giải** bộ nhớ, **không thay đổi dữ liệu và số phần tử**:
  3×4 = 3×2×2 = 2×6 = 12 phần tử.
- Nếu tích các chiều mới khác 12 (ví dụ 5×3 = 15) thì PyTorch báo lỗi.
- Có thể dùng `-1` để PyTorch tự tính một chiều còn lại.


### Bài 6. Phép toán và nhân ma trận
Thực hiện các phép toán phần tử và phép nhân ma trận bằng toán tử `@`.

In [14]:
import torch

a = torch.randn(2, 3)
b = torch.randn(2, 3)

print("a:"); print(a)
print("\nb:"); print(b)
print("\na + b:"); print(a + b)
print("\na / b:"); print(a / b)
print("\na^2:");   print(a ** 2)
print("\nMatrix multiplication (a @ b.T):"); print(a @ b.T)

a:
tensor([[-0.1003,  0.1954, -0.8012],
        [-0.8527, -0.0684, -0.1918]])

b:
tensor([[ 0.9572,  0.4593,  1.1903],
        [ 0.3165, -2.1110,  1.3493]])

a + b:
tensor([[ 0.8570,  0.6547,  0.3891],
        [-0.5362, -2.1794,  1.1575]])

a / b:
tensor([[-0.1048,  0.4253, -0.6731],
        [-2.6938,  0.0324, -0.1421]])

a^2:
tensor([[0.0101, 0.0382, 0.6420],
        [0.7271, 0.0047, 0.0368]])

Matrix multiplication (a @ b.T):
tensor([[-0.9599, -1.5253],
        [-1.0759, -0.3842]])


**Yêu cầu:** Phân biệt phép toán `a * b` với `a @ b.T`; cho biết shape của kết quả `a @ b.T`.

In [15]:
mul_elem = a * b        # nhan tung phan tu (element-wise / Hadamard)
mul_mat  = a @ b.T      # nhan ma tran

print("a * b   (element-wise) shape:", mul_elem.shape)
print(mul_elem)
print("\na @ b.T (matrix mul)  shape:", mul_mat.shape)
print(mul_mat)

# Kiem chung: phan tu (0,0) cua a@b.T chinh la tich vo huong hang 0 cua a va hang 0 cua b
print("\nKiem chung (a@b.T)[0,0] =", (a[0] * b[0]).sum().item(), "vs", mul_mat[0, 0].item())

a * b   (element-wise) shape: torch.Size([2, 3])
tensor([[-0.0960,  0.0897, -0.9537],
        [-0.2699,  0.1444, -0.2587]])

a @ b.T (matrix mul)  shape: torch.Size([2, 2])
tensor([[-0.9599, -1.5253],
        [-1.0759, -0.3842]])

Kiem chung (a@b.T)[0,0] = -0.9599425196647644 vs -0.9599425196647644


**Nhận xét:**
- `a * b` là phép nhân **từng phần tử** (Hadamard). Hai tensor phải cùng shape (hoặc broadcast được);
  kết quả có shape `[2, 3]` giống đầu vào.
- `a @ b.T` là phép **nhân ma trận**: `a` có shape (2, 3), `b.T` có shape (3, 2), điều kiện nhân là số cột của
  ma trận trái bằng số hàng của ma trận phải (3 = 3). Kết quả có shape **`[2, 2]`**.
- Phần tử `(i, j)` của `a @ b.T` là tích vô hướng giữa hàng `i` của `a` và hàng `j` của `b`.
- Không thể viết `a @ b` vì (2,3) @ (2,3) không thoả điều kiện nhân ma trận.


### Bài 7. Các hàm thống kê trên Tensor
Tính trung bình, độ lệch chuẩn và tổng tích lũy theo từng cột.

In [16]:
import torch

a = torch.randn(3, 4)
print(a)
print("\nMean (dim=0):");              print(torch.mean(a, dim=0))
print("\nStandard deviation (dim=0):"); print(torch.std(a, dim=0))
print("\nCumulative sum (dim=0):");     print(torch.cumsum(a, dim=0))

tensor([[ 1.0872,  0.9218,  1.9562,  0.9322],
        [-0.0830, -0.1426, -0.1914, -1.1653],
        [ 1.9960, -0.7709,  0.6036,  1.1748]])

Mean (dim=0):
tensor([1.0001, 0.0027, 0.7895, 0.3139])

Standard deviation (dim=0):
tensor([1.0423, 0.8557, 1.0858, 1.2868])

Cumulative sum (dim=0):
tensor([[ 1.0872,  0.9218,  1.9562,  0.9322],
        [ 1.0041,  0.7792,  1.7648, -0.2331],
        [ 3.0002,  0.0082,  2.3685,  0.9417]])


**Yêu cầu:** Thay `dim=0` bằng `dim=1` và nhận xét sự khác nhau.

In [17]:
print("Mean (dim=1):");               print(torch.mean(a, dim=1))
print("\nStandard deviation (dim=1):"); print(torch.std(a, dim=1))
print("\nCumulative sum (dim=1):");     print(torch.cumsum(a, dim=1))

print("\nShape ket qua:")
print("  mean(dim=0):", torch.mean(a, dim=0).shape, "  mean(dim=1):", torch.mean(a, dim=1).shape)
print("  cumsum(dim=0):", torch.cumsum(a, dim=0).shape, "cumsum(dim=1):", torch.cumsum(a, dim=1).shape)

Mean (dim=1):
tensor([ 1.2243, -0.3956,  0.7509])

Standard deviation (dim=1):
tensor([0.4937, 0.5151, 1.1644])

Cumulative sum (dim=1):
tensor([[ 1.0872,  2.0089,  3.9651,  4.8974],
        [-0.0830, -0.2257, -0.4170, -1.5824],
        [ 1.9960,  1.2251,  1.8288,  3.0035]])

Shape ket qua:
  mean(dim=0): torch.Size([4])   mean(dim=1): torch.Size([3])
  cumsum(dim=0): torch.Size([3, 4]) cumsum(dim=1): torch.Size([3, 4])


**Nhận xét:**
- `dim` là chiều bị **thu gọn (reduce)**. Với tensor `a` shape (3, 4):
  - `dim=0` → gộp theo hàng, tính **theo từng cột**, kết quả là vector 4 phần tử (shape `[4]`).
  - `dim=1` → gộp theo cột, tính **theo từng hàng**, kết quả là vector 3 phần tử (shape `[3]`).
- Riêng `cumsum` không làm giảm số chiều mà chỉ cộng dồn dọc theo chiều được chỉ định, nên kết quả
  vẫn giữ shape `[3, 4]`: `dim=0` cộng dồn từ trên xuống, `dim=1` cộng dồn từ trái sang phải.
- Lưu ý: `torch.std` mặc định dùng ước lượng không chệch (chia cho n-1).


### Bài 8. Autograd cơ bản
Tạo biến có `requires_grad=True` và dùng `backward()` để tính đạo hàm tự động.

In [18]:
import torch

x = torch.tensor(3.6, requires_grad=True)
y = x * x
y.backward()

print("x =", x)
print("y =", y)
print("x.grad =", x.grad)

x = tensor(3.6000, requires_grad=True)
y = tensor(12.9600, grad_fn=<MulBackward0>)
x.grad = tensor(7.2000)


**Yêu cầu 1:** Tính thủ công đạo hàm của y = x² tại x = 3.6 và so sánh với `x.grad`.

In [19]:
manual = 2 * 3.6          # dy/dx = 2x  ->  2 * 3.6 = 7.2
print("Dao ham thu cong dy/dx = 2x = 2 * 3.6 =", manual)
print("Autograd  x.grad       =", x.grad.item())
print("Bang nhau?", abs(manual - x.grad.item()) < 1e-6)

Dao ham thu cong dy/dx = 2x = 2 * 3.6 = 7.2
Autograd  x.grad       = 7.199999809265137
Bang nhau? True


**Yêu cầu 2:** Đổi hàm thành y = 3x² + 2x + 1 và kiểm tra gradient.

In [20]:
x2 = torch.tensor(3.6, requires_grad=True)
y2 = 3 * x2 ** 2 + 2 * x2 + 1
y2.backward()

manual2 = 6 * 3.6 + 2     # dy/dx = 6x + 2
print("y  =", y2.item())
print("Dao ham thu cong dy/dx = 6x + 2 =", manual2)
print("Autograd  x2.grad              =", x2.grad.item())
print("Bang nhau?", abs(manual2 - x2.grad.item()) < 1e-5)

y  = 47.07999801635742
Dao ham thu cong dy/dx = 6x + 2 = 23.6
Autograd  x2.grad              = 23.599998474121094
Bang nhau? True


**Nhận xét:**
- Với y = x², đạo hàm là 2x; tại x = 3.6 cho 7.2 — trùng khớp với `x.grad`.
- Với y = 3x² + 2x + 1, đạo hàm là 6x + 2; tại x = 3.6 cho 23.6 — cũng trùng khớp.
- PyTorch xây dựng **đồ thị tính toán động** khi thực hiện các phép toán trên tensor có
  `requires_grad=True`. Khi gọi `backward()`, nó lan truyền ngược (chain rule) và cộng dồn kết quả vào `.grad`.
- Lưu ý: `.grad` được **cộng dồn**, nên trong vòng lặp huấn luyện phải gọi `zero_grad()` trước mỗi bước.


### Bài 9. Gradient Descent tìm hệ số đa thức
Sinh dữ liệu từ đa thức y = x² + 2x + 3, sau đó dùng NAdam và autograd để tìm lại ba hệ số.

In [21]:
import numpy as np
import torch

polynomial = np.poly1d([1, 2, 3])          # y = x^2 + 2x + 3
N = 20
X = np.random.randn(N, 1) * 5
Y = polynomial(X)
XX = np.hstack([X * X, X, np.ones_like(X)])   # ma tran dac trung [x^2, x, 1]

w = torch.randn(3, 1, requires_grad=True)
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)

print("Initial coefficients:")
print(w)

for epoch in range(1000):
    optimizer.zero_grad()              # xoa gradient cu
    y_pred = x @ w                     # du doan
    mse = torch.mean(torch.square(y - y_pred))   # ham mat mat
    mse.backward()                     # lan truyen nguoc, tinh gradient
    optimizer.step()                   # cap nhat tham so
    if (epoch + 1) % 200 == 0:
        print("Epoch %4d | MSE = %.6f" % (epoch + 1, mse.item()))

print("\nLearned coefficients:")
print(w)

Initial coefficients:
tensor([[ 0.8514],
        [-1.2482],
        [-0.2028]], requires_grad=True)
Epoch  200 | MSE = 53.125343
Epoch  400 | MSE = 7.226749
Epoch  600 | MSE = 0.519083
Epoch  800 | MSE = 0.018923
Epoch 1000 | MSE = 0.000342

Learned coefficients:
tensor([[0.9999],
        [1.9961],
        [2.9937]], requires_grad=True)


**Yêu cầu 1:** So sánh ba hệ số học được với [1, 2, 3].

In [22]:
learned = w.detach().numpy().ravel()
true    = np.array([1.0, 2.0, 3.0])
names   = ["a (he so x^2)", "b (he so x)", "c (hang so)"]

print("%-16s %12s %12s %12s" % ("He so", "That", "Hoc duoc", "Sai so"))
for n, t, l in zip(names, true, learned):
    print("%-16s %12.4f %12.4f %12.6f" % (n, t, l, abs(t - l)))

print("\nMSE cuoi cung:", mse.item())

He so                    That     Hoc duoc       Sai so
a (he so x^2)          1.0000       0.9999     0.000056
b (he so x)            2.0000       1.9961     0.003883
c (hang so)            3.0000       2.9937     0.006258

MSE cuoi cung: 0.00034206840791739523


**Nhận xét:** Ba hệ số học được xấp xỉ [1, 2, 3] với sai số rất nhỏ. Sai số còn lại là do số bước lặp hữu hạn
(1000 epoch) và learning rate; nếu tăng số epoch hoặc điều chỉnh `lr` thì hệ số sẽ càng chính xác.
Vì dữ liệu được sinh ra không có nhiễu và mô hình tuyến tính theo tham số, nghiệm tối ưu chính là [1, 2, 3].

**Yêu cầu 2:** Giải thích vai trò của `optimizer.zero_grad()`, `mse.backward()` và `optimizer.step()`.

**Trả lời:**

| Hàm | Vai trò |
|---|---|
| `optimizer.zero_grad()` | Xoá (đặt về 0) gradient của các tham số từ bước lặp trước. PyTorch **cộng dồn** gradient vào `.grad`, nếu không xoá thì gradient các vòng lặp sẽ bị cộng chồng lên nhau và làm hỏng quá trình huấn luyện. |
| `mse.backward()` | Lan truyền ngược trên đồ thị tính toán, áp dụng quy tắc chuỗi để tính ∂MSE/∂w và lưu vào `w.grad`. |
| `optimizer.step()` | Dùng gradient vừa tính để **cập nhật tham số** theo thuật toán tối ưu (ở đây là NAdam - Adam kết hợp Nesterov momentum). |

Thứ tự bắt buộc là: `zero_grad()` → tính loss → `backward()` → `step()`.


### Bài 10. Giải hệ phương trình bằng Autograd
Tối thiểu hóa tổng sai số bình phương để tìm A, B, C, D thỏa bốn phương trình:
A + B = 9, C − D = 1, A + C = 8, B − D = 2.

In [23]:
import random
import torch

A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)

EPOCHS = 2000
optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)

for epoch in range(EPOCHS):
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2
    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4

    optimizer.zero_grad()
    sqerr.backward()
    optimizer.step()
    if (epoch + 1) % 500 == 0:
        print("Epoch %4d | sqerr = %.8f" % (epoch + 1, sqerr.item()))

print("\nA =", A)
print("B =", B)
print("C =", C)
print("D =", D)

Epoch  500 | sqerr = 4.52070713


Epoch 1000 | sqerr = 0.01878936


Epoch 1500 | sqerr = 0.00000535


Epoch 2000 | sqerr = 0.00000000

A = tensor(4.7772, requires_grad=True)
B = tensor(4.2227, requires_grad=True)
C = tensor(3.2228, requires_grad=True)
D = tensor(2.2227, requires_grad=True)


**Yêu cầu 1:** Kiểm tra nghiệm thu được với bốn phương trình.

In [24]:
a_, b_, c_, d_ = A.item(), B.item(), C.item(), D.item()
print("A = %.6f, B = %.6f, C = %.6f, D = %.6f\n" % (a_, b_, c_, d_))

checks = [("A + B = 9", a_ + b_, 9),
          ("C - D = 1", c_ - d_, 1),
          ("A + C = 8", a_ + c_, 8),
          ("B - D = 2", b_ - d_, 2)]

print("%-12s %12s %10s %12s" % ("Phuong trinh", "Gia tri tinh", "Ve phai", "Sai so"))
for name, val, rhs in checks:
    print("%-12s %12.6f %10d %12.6f" % (name, val, rhs, abs(val - rhs)))

A = 4.777212, B = 4.222733, C = 3.222769, D = 2.222748

Phuong trinh Gia tri tinh    Ve phai       Sai so
A + B = 9        8.999946          9     0.000054
C - D = 1        1.000021          1     0.000021
A + C = 8        7.999981          8     0.000019
B - D = 2        1.999985          2     0.000015


**Yêu cầu 2:** Nhận xét vì sao bài toán có thể có nhiều nghiệm.

**Trả lời:** Hệ có 4 phương trình và 4 ẩn, nhưng các phương trình **không độc lập tuyến tính**. Cụ thể:

(A + B) − (C − D) = 9 − 1 = 8, mà (A + C) − (B − D) = 8 − 2 = 6...

Xét kỹ hơn: từ pt1: B = 9 − A; từ pt3: C = 8 − A; thay vào pt2: D = C − 1 = 7 − A;
kiểm tra pt4: B − D = (9 − A) − (7 − A) = 2, đúng bằng vế phải — pt4 **tự động thoả mãn** với mọi A.

Vậy phương trình thứ 4 là **hệ quả của ba phương trình đầu** (hạng của ma trận hệ số chỉ bằng 3 < 4),
nên hệ có **vô số nghiệm** phụ thuộc một tham số tự do A:
> A = t, B = 9 − t, C = 8 − t, D = 7 − t (với t bất kỳ).

Do đó nghiệm mà autograd tìm được phụ thuộc vào **giá trị khởi tạo ngẫu nhiên**: mỗi lần chạy sẽ cho một
bộ (A, B, C, D) khác nhau, nhưng tất cả đều thoả mãn cả bốn phương trình (sqerr ≈ 0) và đều thoả
A + B = 9, B = 9 − A, C = 8 − A, D = 7 − A.

In [25]:
# Kiem chung: nghiem thu duoc co dung dang (t, 9-t, 8-t, 7-t) khong?
t = A.item()
print("t        =", round(t, 6))
print("9 - t    =", round(9 - t, 6), " | B =", round(B.item(), 6))
print("8 - t    =", round(8 - t, 6), " | C =", round(C.item(), 6))
print("7 - t    =", round(7 - t, 6), " | D =", round(D.item(), 6))

# Chay lai voi khoi tao khac de thay nghiem khac
A2 = torch.tensor(50.0, requires_grad=True)
B2 = torch.tensor(-3.0, requires_grad=True)
C2 = torch.tensor(0.0,  requires_grad=True)
D2 = torch.tensor(10.0, requires_grad=True)
opt2 = torch.optim.NAdam([A2, B2, C2, D2], lr=0.05)
for _ in range(5000):
    e = (A2+B2-9)**2 + (C2-D2-1)**2 + (A2+C2-8)**2 + (B2-D2-2)**2
    opt2.zero_grad(); e.backward(); opt2.step()
print("\nKhoi tao khac -> nghiem khac:")
print("A=%.4f B=%.4f C=%.4f D=%.4f | sqerr=%.8f" % (A2.item(), B2.item(), C2.item(), D2.item(), e.item()))

t        = 4.777212
9 - t    = 4.222788  | B = 4.222733
8 - t    = 3.222788  | C = 3.222769
7 - t    = 2.222788  | D = 2.222748



Khoi tao khac -> nghiem khac:
A=26.9209 B=-17.9209 C=-18.9209 D=-19.9208 | sqerr=0.00000000



## 3. Bài tập áp dụng

### Bài 11. Viết chương trình PyTorch để thực hiện các yêu cầu sau:
- Tạo Tensor X có 24 phần tử và reshape lần lượt thành 4 × 6, 2 × 12 và 2 × 3 × 4.
- Tạo hai Tensor A, B kích thước 3 × 4; tính A+B, A*B, A/B, trung bình theo cột và độ lệch chuẩn theo cột.
- Cho y = 2x² + 5x + 3. Dùng autograd tính dy/dx tại x = 2 và kiểm tra bằng phép tính thủ công.
- Thay đa thức trong Bài 9 bằng y = 2x² - 3x + 5; huấn luyện và cho biết các hệ số tìm được.

**Yêu cầu 1: Tensor X 24 phần tử, reshape 4×6, 2×12, 2×3×4**

In [26]:
import torch

X = torch.arange(24)
print("X:", X)
print("shape:", X.shape, "| numel:", X.numel())

print("\nReshape 4 x 6:")
print(X.reshape(4, 6))
print("\nReshape 2 x 12:")
print(X.reshape(2, 12))
print("\nReshape 2 x 3 x 4:")
print(X.reshape(2, 3, 4))

print("\nKiem tra so phan tu sau moi lan reshape:",
      X.reshape(4, 6).numel(), X.reshape(2, 12).numel(), X.reshape(2, 3, 4).numel())

X: tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23])
shape: torch.Size([24]) | numel: 24

Reshape 4 x 6:
tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])

Reshape 2 x 12:
tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])

Reshape 2 x 3 x 4:
tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])

Kiem tra so phan tu sau moi lan reshape: 24 24 24


**Nhận xét:** Mọi cách reshape đều giữ nguyên 24 phần tử vì 4×6 = 2×12 = 2×3×4 = 24.
Reshape chỉ thay đổi cách "nhìn" dữ liệu chứ không sao chép hay thay đổi giá trị.

**Yêu cầu 2: Hai Tensor A, B kích thước 3 × 4 — tính A+B, A\*B, A/B, trung bình và độ lệch chuẩn theo cột**

In [27]:
import torch

A = torch.randn(3, 4)
B = torch.randn(3, 4)

print("A:"); print(A)
print("\nB:"); print(B)
print("\nA + B:"); print(A + B)
print("\nA * B (element-wise):"); print(A * B)
print("\nA / B:"); print(A / B)
print("\nTrung binh theo cot  - mean(dim=0):"); print(torch.mean(A, dim=0))
print("\nDo lech chuan theo cot - std(dim=0):"); print(torch.std(A, dim=0))

A:
tensor([[-0.0861, -0.7230, -0.0691,  0.9317],
        [-0.0499,  1.0113, -1.3453,  1.4902],
        [-0.5966,  0.7638, -1.9679,  0.8976]])

B:
tensor([[-0.3688, -0.7289, -0.3883, -0.5032],
        [ 0.4745, -1.0572, -1.6563,  1.4061],
        [ 0.0867,  1.3332, -0.0777,  1.6055]])

A + B:
tensor([[-0.4549, -1.4518, -0.4575,  0.4285],
        [ 0.4245, -0.0459, -3.0016,  2.8963],
        [-0.5098,  2.0970, -2.0456,  2.5031]])

A * B (element-wise):
tensor([[ 0.0317,  0.5269,  0.0268, -0.4689],
        [-0.0237, -1.0691,  2.2281,  2.0954],
        [-0.0517,  1.0183,  0.1530,  1.4411]])

A / B:
tensor([[ 0.2334,  0.9919,  0.1780, -1.8516],
        [-0.1053, -0.9566,  0.8122,  1.0598],
        [-6.8796,  0.5729, 25.3164,  0.5591]])

Trung binh theo cot  - mean(dim=0):
tensor([-0.2442,  0.3507, -1.1274,  1.1065])

Do lech chuan theo cot - std(dim=0):
tensor([0.3057, 0.9380, 0.9679, 0.3327])


**Nhận xét:** `A+B`, `A*B`, `A/B` đều là phép toán theo từng phần tử nên kết quả giữ shape `[3, 4]`.
`mean(dim=0)` và `std(dim=0)` thu gọn chiều hàng nên trả về vector 4 phần tử — mỗi phần tử ứng với một cột.

**Yêu cầu 3: Cho y = 2x² + 5x + 3 — dùng autograd tính dy/dx tại x = 2 và kiểm tra bằng tính thủ công**

In [28]:
import torch

x = torch.tensor(2.0, requires_grad=True)
y = 2 * x**2 + 5 * x + 3
y.backward()

manual = 4 * 2 + 5        # dy/dx = 4x + 5
print("y(2)                       =", y.item())
print("Autograd  dy/dx tai x = 2  =", x.grad.item())
print("Thu cong  dy/dx = 4x + 5   =", manual)
print("Bang nhau?", abs(manual - x.grad.item()) < 1e-6)

y(2)                       = 21.0
Autograd  dy/dx tai x = 2  = 13.0
Thu cong  dy/dx = 4x + 5   = 13
Bang nhau? True


**Nhận xét:** y = 2x² + 5x + 3 ⟹ dy/dx = 4x + 5. Tại x = 2: dy/dx = 4(2) + 5 = **13**,
đúng bằng giá trị `x.grad` do autograd tính ra.

**Yêu cầu 4: Thay đa thức trong Bài 9 bằng y = 2x² − 3x + 5 — huấn luyện và cho biết các hệ số tìm được**

In [29]:
import numpy as np
import torch

polynomial = np.poly1d([2, -3, 5])        # y = 2x^2 - 3x + 5
N = 20
X = np.random.randn(N, 1) * 5
Y = polynomial(X)
XX = np.hstack([X * X, X, np.ones_like(X)])

w = torch.randn(3, 1, requires_grad=True)
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y,  dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)
print("Initial coefficients:", w.detach().numpy().ravel())

for epoch in range(5000):
    optimizer.zero_grad()
    y_pred = x @ w
    mse = torch.mean(torch.square(y - y_pred))
    mse.backward()
    optimizer.step()
    if (epoch + 1) % 1000 == 0:
        print("Epoch %5d | MSE = %.8f" % (epoch + 1, mse.item()))

learned = w.detach().numpy().ravel()
true    = np.array([2.0, -3.0, 5.0])
print("\n%-16s %12s %12s %12s" % ("He so", "That", "Hoc duoc", "Sai so"))
for n, t, l in zip(["a (x^2)", "b (x)", "c (hang so)"], true, learned):
    print("%-16s %12.4f %12.4f %12.6f" % (n, t, l, abs(t - l)))

Initial coefficients: [-0.60293436  1.5979165   0.06389374]
Epoch  1000 | MSE = 6.32188702


Epoch  2000 | MSE = 0.87618768
Epoch  3000 | MSE = 0.21018200


Epoch  4000 | MSE = 0.01863561
Epoch  5000 | MSE = 0.00032257

He so                    That     Hoc duoc       Sai so
a (x^2)                2.0000       2.0004     0.000400
b (x)                 -3.0000      -3.0005     0.000549
c (hang so)            5.0000       4.9764     0.023646


**Nhận xét:** Sau 5000 epoch, ba hệ số học được hội tụ rất gần [2, −3, 5] và MSE giảm về xấp xỉ 0.
So với Bài 9, ở đây cần nhiều epoch hơn vì hệ số cần học có biên độ lớn hơn và xuất phát điểm ngẫu nhiên
có thể xa nghiệm hơn. Điều này minh hoạ: cùng một mô hình tuyến tính theo tham số, Gradient Descent + autograd
có thể khôi phục lại chính xác các hệ số của đa thức sinh dữ liệu.